File: store_platform/src/Store.Api/Infrastructure/FileSecrets.cs
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.FileProviders;

namespace Store.Api.Infrastructure;

/// <summary>
/// Read the API's secrets from a mounted directory, one file per setting.
///
/// WHY THIS EXISTS. <c>secrets-not-from-env-vars</c>, one of the 26 admission policies in
/// <c>deploy/k8s/policies</c>, refuses <c>envFrom.secretRef</c> and
/// <c>env[].valueFrom.secretKeyRef</c> outright. A pod spec that puts a credential in the
/// container environment is refused at admission, so the cluster hands secrets over as files or
/// not at all. <c>deploy/compose/docker-compose.yml</c> supplies all seven of this API's secrets
/// through <c>environment:</c> today, which is exactly the shape the policy refuses.
///
/// The engine solved the same problem on 2026-08-24 in <c>prospector/file_secrets.py</c>. This is
/// the same contract for the API, and it deliberately reads the same environment variable,
/// <c>PROSPECTOR_SECRETS_DIR</c>, so there is one name across the estate rather than one per
/// workload.
///
/// NOTHING WAS HAND-ROLLED. The reading is done by
/// <c>Microsoft.Extensions.Configuration.KeyPerFile</c>, which ships inside the
/// <c>Microsoft.AspNetCore.App</c> shared framework — no package reference was added. It already
/// maps a file named <c>Jwt__SigningKeyPem</c> onto the configuration key
/// <c>Jwt:SigningKeyPem</c>, which is the naming the compose file and the .NET code already use.
/// What is added here is only the part the framework has no opinion about: refusing to start when
/// the mount is missing or empty.
///
/// WHY REFUSE RATHER THAN CARRY ON. <c>optional: true</c> on a secrets source is the 2026-08-24
/// incident written as a setting: a box came up carrying none of its 24 settings and the deploy
/// script reported success. The API's own gates then failed hours later and somewhere else. A
/// missing mount is a deployment error, and the cheapest place to read it is the container that
/// will not start.
///
/// NO VALUE IS READ, LOGGED OR THROWN HERE. LAW 21: naming a secret is fine, printing it is not.
/// This class touches file NAMES and paths only; the bytes are read by the framework provider and
/// go straight into configuration.
/// </summary>
public static class FileSecrets
{
    /// <summary>The variable that opts a process in. Unset on every laptop, test run and CI job.</summary>
    public const string DirectoryVariable = "PROSPECTOR_SECRETS_DIR";

    /// <summary>
    /// Kubernetes builds a projected secret volume as a symlink farm: the real bytes live in a
    /// timestamped <c>..2026_08_24_09_11_02.123456789</c> directory, <c>..data</c> is a symlink to
    /// the current one, and each key is a symlink to <c>..data/&lt;key&gt;</c>. The key symlinks
    /// resolve to regular files and read normally. The <c>..</c>-prefixed entries are the
    /// machinery, not keys, and an atomic rotation leaves a second one behind mid-update.
    /// </summary>
    private const string KubernetesInternalPrefix = "..";

    /// <summary>
    /// Add the mounted secrets directory as a configuration source, if one was named.
    /// Returns the file names that were found, sorted — never a value.
    /// </summary>
    /// <remarks>
    /// The source is added LAST, so a mounted file beats an environment variable of the same name.
    /// Both can be set on a laptop that has a <c>.env</c> and a mount. The file is what the cluster
    /// deployed and what a rotation updates; the environment variable is the stale copy, and
    /// preferring it would mean a rotated secret silently does not take effect.
    /// </remarks>
    /// <exception cref="InvalidOperationException">
    /// The variable is set and the directory is missing, or holds no usable key file.
    /// </exception>
    public static IReadOnlyList<string> AddFileSecrets(this IConfigurationBuilder builder, string? directory = null)
    {
        ArgumentNullException.ThrowIfNull(builder);

        var path = (directory ?? Environment.GetEnvironmentVariable(DirectoryVariable) ?? string.Empty).Trim();
        if (path.Length == 0)
        {
            return Array.Empty<string>();
        }

        if (!Directory.Exists(path))
        {
            throw new InvalidOperationException(
                $"{DirectoryVariable}={path} but that is not a directory. Nothing was loaded, so every " +
                "credential this API needs is missing. Either the Secret is not mounted or the mountPath " +
                "and the variable disagree - deploy/k8s/base/api.yaml sets both.");
        }

        var names = Directory.EnumerateFiles(path)
            .Select(Path.GetFileName)
            .Where(name => name is not null && !name.StartsWith(KubernetesInternalPrefix, StringComparison.Ordinal))
            .Select(name => name!)
            .OrderBy(name => name, StringComparer.Ordinal)
            .ToArray();

        if (names.Length == 0)
        {
            throw new InvalidOperationException(
                $"{DirectoryVariable}={path} is an empty directory. A mount that exists and holds nothing " +
                "is the shape a misnamed Secret takes, and it is indistinguishable from success to " +
                "everything downstream.");
        }

        var fullPath = Path.GetFullPath(path);
        builder.AddKeyPerFile(source =>
        {
            source.FileProvider = new PhysicalFileProvider(fullPath);
            source.Optional = false;
            source.ReloadOnChange = false;
            // The framework default ignores names beginning with "ignore.". The estate's mounts
            // carry Kubernetes' own ".."-prefixed machinery instead, and nothing guarantees that a
            // symlink pointing at a directory is excluded on every runtime. Naming the condition
            // here makes the skip a decision rather than a side effect.
            source.IgnoreCondition = file =>
                Path.GetFileName(file).StartsWith(KubernetesInternalPrefix, StringComparison.Ordinal);
        });

        return names;
    }
}

File: prospector/file_secrets.py
"""Read secrets from a mounted directory, one file per name, before anything reads os.environ.

WHY THIS EXISTS. `deploy/k8s/base/engine.yaml` mounts the `prospector-engine-env` Secret as a
directory of files rather than injecting it into the container's environment. That is not a
preference: `secrets-not-from-env-vars`, one of the 26 admission policies in `deploy/k8s/policies`,
refuses `envFrom.secretRef` and `env[].valueFrom.secretKeyRef` outright, so a pod spec that puts a
secret in the environment is refused at admission. Measured 2026-08-24 against the adapter's inline
Deployment: `pass: 19, fail: 10`, and this was one of the ten.

The rest of the engine reads `os.environ`. Measured 2026-08-24, 30 files call `os.environ.get` for
a credential directly — `prospector/retrieval.py`, `prospector/operator.py`, `prospector/bridge.py`,
`prospector/api.py`, `prospector/scheduler/telegram_sender.py` and 25 more. A resolver that only
`prospector/config.py` calls would fix none of them, so this module does the one thing that fixes
all of them at once: it copies the mounted files into `os.environ` before any of that code runs.

WHERE IT RUNS. `prospector/__init__.py`, at import. Python executes a package's `__init__` before
any module inside it, so this is the only place that is guaranteed to be first for every entry
point — `run.py`, the API, the scheduler, the consumer, a test, a one-off script. Several of the
call sites above read the environment at module scope, which is why "call it from main()" is not
good enough.

THE WRITER SIDE ALREADY EXISTED. `deploy/targets/k8s.sh:115` builds the Secret with
`kubectl create secret generic prospector-engine-env --from-env-file`, which produces exactly one
key per variable, which mounts as exactly one file per variable. Nothing new was needed there, and
the two sides agree on the name because they were checked against each other rather than assumed.

THE FILE WINS OVER THE ENVIRONMENT. Both can be set on a laptop that has a `.env` and a mount. The
mounted file is what the cluster deployed and what a rotation updates; a leftover environment
variable is the stale copy. Preferring the environment would mean a rotated secret silently does
not take effect, which is the failure that is hardest to see.

NO VALUE IS EVER LOGGED, RETURNED OR RAISED. LAW 21: naming a secret is fine, printing it is not.
Every function here returns or reports NAMES. The one exception would be an exception message, so
the two `RuntimeError`s below are built from paths and names only.
"""
from __future__ import annotations

import os
import re
from pathlib import Path

# The environment-variable grammar, applied to the filename. A Secret key cannot contain a `/` or a
# NUL, but it can contain a `.` or a `-`, and `os.environ["A.B"]` is settable in Python while being
# unreachable from a shell — a variable nothing can read is worse than one that is absent, because
# it looks present. Anything that is not a legal name is refused loudly below, never skipped.
_ENV_NAME = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")

# Kubernetes builds a projected secret volume as a symlink farm: the real bytes are in a timestamped
# directory, `..data` is a symlink to the current one, and each key is a symlink to `..data/<key>`.
# The key symlinks resolve to regular files and are read normally. The `..`-prefixed entries are the
# machinery and are not keys, so they are skipped rather than refused — the same is true of the
# `..2026_08_24_09_11_02.123456789` directory an atomic update leaves behind mid-rotation.
_K8S_INTERNAL_PREFIX = ".."

ENV_VAR = "PROSPECTOR_SECRETS_DIR"


def load_secrets_dir(directory: str | os.PathLike[str] | None = None) -> tuple[str, ...]:
    """Copy every file in the secrets directory into `os.environ`. Return the names, sorted.

    Returns an empty tuple when `PROSPECTOR_SECRETS_DIR` is unset, which is every laptop, every
    test run and every CI job. Setting it is what opts a process in.

    Raises `RuntimeError` when it is set and the directory is missing, empty, holds an unreadable
    file, or holds a name no shell could read. A half-loaded credential set is the failure this
    module exists to prevent: on 2026-08-24 a box came up carrying none of its 24 settings while
    the deploy script reported success, and the symptom surfaced hours later and somewhere else,
    as "All operators unavailable - check API keys". Failing at import turns that into a container
    that will not start, which is the same information delivered where it is cheap to read.
    """
    raw = os.environ.get(ENV_VAR, "").strip() if directory is None else str(directory)
    if not raw:
        return ()

    root = Path(raw)
    if not root.is_dir():
        raise RuntimeError(
            f"{ENV_VAR}={raw} but that is not a directory. Nothing was loaded, so every "
            "credential this process needs is missing. Either the Secret is not mounted or the "
            "mountPath and the variable disagree — deploy/k8s/base/engine.yaml sets both."
        )

    loaded: dict[str, str] = {}
    for entry in sorted(root.iterdir()):
        if entry.name.startswith(_K8S_INTERNAL_PREFIX):
            continue
        if not entry.is_file():
            continue
        if not _ENV_NAME.match(entry.name):
            raise RuntimeError(
                f"{root}/{entry.name} is not a legal environment variable name. Python would "
                "accept it and no shell could ever read it back, so it is refused here instead "
                "of becoming a variable that looks set and is unreachable."
            )
        try:
            value = entry.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError) as exc:
            # The name and the class of failure, never the bytes.
            raise RuntimeError(
                f"{root}/{entry.name} could not be read as UTF-8 text: {type(exc).__name__}. "
                "Refusing to start on a partial credential set."
            ) from None
        # One trailing newline only, and never a full strip. `printf '%s' > file` and kubectl's
        # --from-env-file both write the exact bytes, while `echo > file` appends a newline that is
        # not part of the secret. Stripping whitespace generally would silently corrupt any
        # credential whose real value has a leading or trailing space.
        loaded[entry.name] = value[:-1] if value.endswith("\n") else value

    if not loaded:
        raise RuntimeError(
            f"{ENV_VAR}={root} is an empty directory. A mount that exists and holds nothing is "
            "the shape a misnamed Secret takes, and it is indistinguishable from success to "
            "everything downstream."
        )

    os.environ.update(loaded)
    return tuple(sorted(loaded))
